core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1647,6 +1825,36 @@ class KafkaController(val config: KafkaConfig,
     }
   }
 
+  private def processUpdateFinalizedFeatures(newFeatures: Features[FinalizedVersionRange],
+                                             callback: UpdateFinalizedFeaturesCallback): Unit = {
+    if (isActive) {
+      val incompatibleBrokers = controllerContext.liveOrShuttingDownBrokers.filter(broker => {
+        BrokerFeatures.hasIncompatibleFeatures(broker.features, newFeatures)
+      })
+      if (incompatibleBrokers.size > 0) {
+        callback(
+          Errors.INVALID_REQUEST,
+          Some(
+            s"Could not apply finalized feature updates because ${incompatibleBrokers.size} brokers"
+            + s" were found to have incompatible features. newFeatures: $newFeatures"
+           + s", incompatibleBrokers: $incompatibleBrokers."))
+      } else {
+        try {
+          val newVersion = zkClient.updateFeatureZNode(
+            new FeatureZNode(FeatureZNodeStatus.Enabled, newFeatures))
+          featureCache.waitUntilEpochOrThrow(newVersion, config.zkConnectionTimeoutMs)
```

Comment:
Could we avoid blocking controller processing here, by putting the callback into a delayed queue or sth?